File: OTEAServer/Controllers/UsersController.cs
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;
using Microsoft.EntityFrameworkCore;
using Microsoft.IdentityModel.Tokens;
using OTEAServer.Misc;
using OTEAServer.Models;
using System.IdentityModel.Tokens.Jwt;
using System.Net;
using System.Security.Claims;
using System.Text;
using System.Text.Json;

namespace OTEAServer.Controllers
{
    /// <summary>
    /// Controller class for users operations
    /// Author: Pablo Ahíta del Barrio
    /// Version: 1
    /// </summary>
    [ApiController]
    [Route("Users")]
    [Authorize]
    public class UsersController : ControllerBase
    {
        /// <summary>
        /// Database context
        /// </summary>
        private readonly DatabaseContext _context;

        /// <summary>
        /// Session Configuration
        /// </summary>
        private readonly SessionConfig _sessionConfig;

        /// <summary>
        /// Class constructor
        /// </summary>
        /// <param name="context">Database context</param>
        public UsersController(DatabaseContext context, SessionConfig sessionConfig)
        {
            _context = context;
            _sessionConfig = sessionConfig;
        }

        /// <summary>
        /// Method that obtains all the users
        /// </summary>
        /// <returns>User list</returns>
        [HttpGet("allRequests")]
        [Authorize(Policy = "Administrator")]
        public IActionResult GetAllRequests([FromHeader] string Authorization)
        {
            try
            {
                var users = _context.Users.Where(u=>u.isActive==-1).ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach (var user in users)
                {
                    String usr = "{\"emailUser\":\"" + user.emailUser + "\"," +
                    "\"userType\":\"" + user.userType + "\"," +
                        "\"first_name\":\"" + user.first_name + "\"," +
                        "\"last_name\":\"" + user.last_name + "\"," +
                        "\"telephone\":\"" + user.telephone + "\"," +
                        "\"idOrganization\":" + user.idOrganization + "," +
                        "\"orgType\":\"" + user.orgType + "\"," +
                        "\"illness\":\"" + user.illness + "\"," +
                        "\"profilePhoto\":\"" + user.profilePhoto + "\"," +
                        "\"isActive\":\"" + user.isActive + "\"}";
                    result.Add(JsonDocument.Parse(usr));
                }
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("allActiveUsers")]
        [Authorize(Policy = "Administrator")]
        public IActionResult GetAllActiveUsers([FromHeader] string Authorization)
        {
            try
            {
                var users = _context.Users.Where(u => u.isActive == 1 && u.userType == "ORGANIZATION").ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach (var user in users)
                {
                    String usr = "{\"emailUser\":\"" + user.emailUser + "\"," +
                    "\"userType\":\"" + user.userType + "\"," +
                        "\"first_name\":\"" + user.first_name + "\"," +
                        "\"last_name\":\"" + user.last_name + "\"," +
                        "\"telephone\":\"" + user.telephone + "\"," +
                        "\"idOrganization\":" + user.idOrganization + "," +
                        "\"orgType\":\"" + user.orgType + "\"," +
                        "\"illness\":\"" + user.illness + "\"," +
                        "\"profilePhoto\":\"" + user.profilePhoto + "\"," +
                        "\"isActive\":\"" + user.isActive + "\"}";
                    result.Add(JsonDocument.Parse(usr));
                }
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }




        /// <summary>
        /// Method that obtains all the users
        /// </summary>
        /// <returns>User list</returns>
        [HttpGet("all")]
        public IActionResult GetAll([FromHeader] string Authorization)
        {
            try
            {
                var users = _context.Users.Where(u=>u.isActive==1).ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach(var user in users) {
                    String usr = "{\"emailUser\":\"" + user.emailUser + "\"," +
                    "\"userType\":\"" + user.userType + "\"," +
                        "\"first_name\":\"" + user.first_name + "\"," +
                        "\"last_name\":\"" + user.last_name + "\"," +
                        "\"telephone\":\"" + user.telephone + "\"," +
                        "\"idOrganization\":" + user.idOrganization + "," +
                        "\"orgType\":\"" + user.orgType + "\"," +
                        "\"illness\":\"" + user.illness + "\"," +
                        "\"profilePhoto\":\"" + user.profilePhoto + "\"," +
                        "\"isActive\":\"" + user.isActive + "\"}";
                    result.Add(JsonDocument.Parse(usr));
                }
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }


        /// <summary>
        /// Method that obtains all organization users of an organization
        /// </summary>
        /// <param name="idOrganization">User organization identifier</param>
        /// <param name="orgType">User organization type</param>
        /// <param name="illness">User organization illness or syndrome</param>
        /// <returns>User list</returns>
        [HttpGet("allByOrg")]
        public IActionResult GetAllOrgUsersByOrganization([FromQuery] int? idOrganization, [FromQuery] string? orgType, [FromQuery] string? illness, [FromHeader] string Authorization)
        {
            try
            {
                var users = _context.Users.Where(u => u.idOrganization == idOrganization && u.orgType == orgType && u.illness == illness && u.isActive == 1).ToList();
                List<JsonDocument> result = new List<JsonDocument>();
                foreach (var user in users)
                {
                    String usr = "{\"emailUser\":\"" + user.emailUser + "\"," +
                    "\"userType\":\"" + user.userType + "\"," +
                        "\"first_name\":\"" + user.first_name + "\"," +
                        "\"last_name\":\"" + user.last_name + "\"," +
                        "\"telephone\":\"" + user.telephone + "\"," +
                        "\"idOrganization\":" + user.idOrganization + "," +
                        "\"orgType\":\"" + user.orgType + "\"," +
                        "\"illness\":\"" + user.illness + "\"," +
                        "\"profilePhoto\":\"" + user.profilePhoto + "\"," +
                        "\"isActive\":\"" + user.isActive + "\"}";
                    result.Add(JsonDocument.Parse(usr));
                }
                return Ok(result);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains an user from database
        /// </summary>
        /// <param name="email">User email</param>
        /// <returns>User if success, null if not</returns>
        [HttpGet("get")]
        [AllowAnonymous]
        public ActionResult<JsonDocument> Get([FromQuery] string email)
        {
            try
            {
                var user = _context.Users.FirstOrDefault(u => u.emailUser == email);

                if (user == null)
                    return NotFound();
                String usr = "{\"emailUser\":\"" + user.emailUser + "\"," +
                    "\"userType\":\"" + user.userType + "\"," +
                        "\"first_name\":\"" + user.first_name + "\"," +
                        "\"last_name\":\"" + user.last_name + "\"," +
                        "\"telephone\":\"" + user.telephone + "\"," +
                        "\"idOrganization\":" + user.idOrganization + "," +
                        "\"orgType\":\"" + user.orgType + "\"," +
                        "\"illness\":\"" + user.illness + "\"," +
                        "\"profilePhoto\":\"" + user.profilePhoto + "\"," +
                        "\"isActive\":\"" + user.isActive + "\"}";
                return JsonDocument.Parse(usr);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        [HttpGet("getDirector")]
        [Authorize]
        public ActionResult<JsonDocument> GetDirector([FromQuery] int idOrganization, [FromQuery] string organizationType, [FromQuery] string illness, [FromHeader] string Authorization)
        {
            try
            {
                var user = _context.Users.FirstOrDefault(u => u.idOrganization==idOrganization && u.orgType==organizationType && u.illness==illness && u.userType=="DIRECTOR");

                if (user == null)
                    return NotFound();
                String usr = "{\"emailUser\":\"" + user.emailUser + "\"," +
                    "\"userType\":\"" + user.userType + "\"," +
                        "\"first_name\":\"" + user.first_name + "\"," +
                        "\"last_name\":\"" + user.last_name + "\"," +
                        "\"telephone\":\"" + user.telephone + "\"," +
                        "\"idOrganization\":" + user.idOrganization + "," +
                        "\"orgType\":\"" + user.orgType + "\"," +
                        "\"illness\":\"" + user.illness + "\"," +
                        "\"profilePhoto\":\"" + user.profilePhoto + "\"," +
                        "\"isActive\":\"" + user.isActive + "\"}";
                return JsonDocument.Parse(usr);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains the login
        /// </summary>
        /// <param name="email">User login</param>
        /// <param name="password">User password</param>
        /// <returns>Login if credentials are true, null if not</returns>
        [HttpPost("login")]
        [AllowAnonymous]
        public ActionResult Login([FromBody] JsonDocument credentials)
        {
            try
            {
                string email=credentials.RootElement.GetProperty("email").ToString();
                string password = credentials.RootElement.GetProperty("password").ToString();
                var user = _context.Users.FirstOrDefault(u => u.emailUser == email && u.passwordUser == password && u.isActive == 1);

                if (user == null)
                {//Finds user in database
                    user = _context.Users.FirstOrDefault(u => u.emailUser == email);
                    if (user == null) { return NotFound(); }
                    else
                    {
                        if (user.passwordUser != password)
                        {
                            return Unauthorized();
                        }
                    }
                }


                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes(_sessionConfig.secret);
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(new Claim[]
                    {
                        new Claim(ClaimTypes.Name, user.emailUser),
                        new Claim(ClaimTypes.Role, user.userType)
                    }),
                    Expires = DateTime.UtcNow.AddHours(12),
                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                string sessionToken = tokenHandler.WriteToken(token);
                var usr = new
                {
                    emailUser = email,
                    userType = user.userType,
                    first_name = user.first_name,
                    last_name = user.last_name,
                    telephone = user.telephone,
                    idOrganization = user.idOrganization,
                    orgType = user.orgType,
                    illness = user.illness,
                    profilePhoto = user.profilePhoto,
                    isActive = user.isActive
                };
                var organization= _context.Organizations.FirstOrDefault(o => o.idOrganization == user.idOrganization && o.orgType == user.orgType && o.illness == user.illness);
                if (organization == null) {
                    return BadRequest();
                }
                var org = new {
                    idOrganization = organization.idOrganization,
                    orgType = organization.orgType,
                    illness = organization.illness,
                    nameOrg = organization.nameOrg,
                    idAddress = organization.idAddress,
                    email = organization.email,
                    telephone = organization.telephone,
                    informationSpanish = organization.informationSpanish,
                    informationEnglish = organization.informationEnglish,
                    informationFrench = organization.informationFrench,
                    informationBasque = organization.informationBasque,
                    informationCatalan = organization.informationCatalan,
                    informationDutch = organization.informationDutch,
                    informationGalician = organization.informationGalician,
                    informationGerman = organization.informationGerman,
                    informationItalian = organization.informationItalian,
                    informationPortuguese = organization.informationPortuguese,
                    profilePhoto = organization.profilePhoto
                };
                var response = new
                {
                    user = usr,
                    organization = org,
                    token = "Bearer "+sessionToken

                };
                return Ok(response);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that obtains the login
        /// </summary>
        /// <param name="email">User login</param>
        /// <param name="password">User password</param>
        /// <returns>Login if credentials are true, null if not</returns>
        [HttpPost("loginAdmin")]
        [AllowAnonymous]
        public ActionResult LoginAdmin([FromBody] JsonDocument credentials)
        {
            try
            {
                string email = credentials.RootElement.GetProperty("email").ToString();
                string password = credentials.RootElement.GetProperty("password").ToString();
                var user = _context.Users.FirstOrDefault(u => u.emailUser == email && u.passwordUser == password && u.userType=="ADMIN" && u.isActive == 1);

                if (user == null)
                {//Finds user in database
                    user = _context.Users.FirstOrDefault(u => u.emailUser == email && u.userType == "ADMIN" && u.isActive == 1);
                    if (user == null) { return NotFound(); }
                    else
                    {
                        if (user.passwordUser != password)
                        {
                            return Unauthorized();
                        }
                    }
                }


                var tokenHandler = new JwtSecurityTokenHandler();
                var key = Encoding.ASCII.GetBytes(_sessionConfig.secret);
                var tokenDescriptor = new SecurityTokenDescriptor
                {
                    Subject = new ClaimsIdentity(new Claim[]
                    {
                        new Claim(ClaimTypes.Name, user.emailUser),
                        new Claim(ClaimTypes.Role, user.userType)
                    }),
                    Expires = DateTime.UtcNow.AddHours(12),
                    SigningCredentials = new SigningCredentials(new SymmetricSecurityKey(key), SecurityAlgorithms.HmacSha256Signature)
                };
                var token = tokenHandler.CreateToken(tokenDescriptor);
                string sessionToken = tokenHandler.WriteToken(token);
                var usr = new
                {
                    emailUser = email,
                    userType = user.userType,
                    first_name = user.first_name,
                    last_name = user.last_name,
                    telephone = user.telephone,
                    idOrganization = user.idOrganization,
                    orgType = user.orgType,
                    illness = user.illness,
                    profilePhoto = user.profilePhoto,
                    isActive = user.isActive
                };
                var organization = _context.Organizations.FirstOrDefault(o => o.idOrganization == user.idOrganization && o.orgType == user.orgType && o.illness == user.illness);
                if (organization == null)
                {
                    return BadRequest();
                }
                var org = new
                {
                    idOrganization = organization.idOrganization,
                    orgType = organization.orgType,
                    illness = organization.illness,
                    nameOrg = organization.nameOrg,
                    idAddress = organization.idAddress,
                    email = organization.email,
                    telephone = organization.telephone,
                    informationSpanish = organization.informationSpanish,
                    informationEnglish = organization.informationEnglish,
                    informationFrench = organization.informationFrench,
                    informationBasque = organization.informationBasque,
                    informationCatalan = organization.informationCatalan,
                    informationDutch = organization.informationDutch,
                    informationGalician = organization.informationGalician,
                    informationGerman = organization.informationGerman,
                    informationItalian = organization.informationItalian,
                    informationPortuguese = organization.informationPortuguese,
                    profilePhoto = organization.profilePhoto
                };
                var response = new
                {
                    user = usr,
                    organization = org,
                    token = "Bearer " + sessionToken

                };
                return Ok(response);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that appends an user to the database
        /// </summary>
        /// <param name="user">User</param>
        /// <returns>User if success, null if not</returns>
        [HttpPost]
        [AllowAnonymous]
        public IActionResult Create([FromBody] User user)
        {
            try
            {
                _context.Users.Add(user);
                _context.SaveChanges();
                return CreatedAtAction(nameof(Get), new { email = user.emailUser }, user);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that updates an user
        /// </summary>
        /// <param name="email">User email</param>
        /// <param name="user">User</param>
        /// <returns>User if success, null if not</returns>
        [HttpPut]
        [AllowAnonymous]
        public IActionResult Update([FromQuery] string email, [FromBody] User user)
        {
            try
            {
                if (email != user.emailUser)
                    return BadRequest();

                var existingUser = _context.Users.FirstOrDefault(u => u.emailUser == email);
                if (existingUser is null)
                    return NotFound();

                existingUser.emailUser = email;
                existingUser.userType = user.userType;
                existingUser.first_name = user.first_name;
                existingUser.last_name = user.last_name;
                if (!user.passwordUser.IsNullOrEmpty()) {
                    existingUser.passwordUser = user.passwordUser;
                }
                existingUser.telephone = user.telephone;
                existingUser.idOrganization = user.idOrganization;
                existingUser.orgType = user.orgType;
                existingUser.illness = user.illness;
                existingUser.profilePhoto = user.profilePhoto;
                existingUser.isActive = user.isActive;
                _context.SaveChanges();

                return Ok(existingUser);
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }

        /// <summary>
        /// Method that deletes an user
        /// </summary>
        /// <param name="email">User email</param>
        /// <returns>User if success, null if not</returns>
        [HttpDelete]
        [Authorize(Policy = "Administrator")]
        public IActionResult Delete([FromQuery] string email, [FromHeader] string Authorization)
        {
            try
            {
                var user = _context.Users.FirstOrDefault(u => u.emailUser == email);

                if (user is null)
                    return NotFound();

                if (user.isActive == 1)
                {
                    var director = _context.Users.FirstOrDefault(u => u.idOrganization == user.idOrganization && u.orgType == user.orgType && u.illness == user.illness && u.userType == "DIRECTOR");
                    if (director is null)
                        return NotFound();
                    var evaluatorTeams = _context.EvaluatorTeams.Where(e => e.emailProfessional == email || e.emailResponsible == email);
                    if (evaluatorTeams.Count() > 0) {
                        foreach (var evaluatorTeam in evaluatorTeams) {
                            if (evaluatorTeam.emailProfessional == email) {
                                evaluatorTeam.emailProfessional = director.emailUser;
                            }
                            if (evaluatorTeam.emailResponsible == email)
                            {
                                evaluatorTeam.emailResponsible = director.emailUser;
                            }
                        }
                    }
                    _context.SaveChanges();
                }
                

                _context.Users.Remove(user);
                _context.SaveChanges();

                return NoContent();
            }
            catch (Exception ex)
            {
                return BadRequest(ex.Message);
            }
        }
    }
}

File: connection/src/main/java/otea/connection/controller/UsersController.java
package otea.connection.controller;



import java.io.IOException;
import java.net.SocketTimeoutException;
import java.util.ArrayList;
import java.util.List;
import java.util.concurrent.Callable;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;

import cli.organization.Organization;
import cli.organization.data.EvaluatorTeam;
import cli.user.User;
import otea.connection.ConnectionClient;
import otea.connection.api.UsersApi;
import retrofit2.Call;
import retrofit2.Response;
import session.Session;

import com.google.gson.JsonObject;

/**
 * Controller class for users operations
 *
 * @author Pablo Ahíta del Barrio
 * @version 1
 * */
public class UsersController {

    /**Users api to connect to the server*/
    private static UsersApi api;

    /**Controller instance*/
    private static UsersController instance;

    /**Number of attempts*/
    private static int numAttempts=0;

    /**Class controller*/
    private UsersController(){
        api= ConnectionClient.getInstance().getRetrofit().create(UsersApi.class);
    }

    /**
     * Method that obtains the singleton instance of the controller
     *
     * @return Controller instance
     * */
    public static UsersController getInstance(){
        if(instance==null){
            synchronized (UsersController.class){
                if(instance==null){
                    instance=new UsersController();
                }
            }
        }
        return instance;
    }

    /**Refresh API*/
    public static void refreshApi(){
        instance=new UsersController();
    }

    public interface LoginCallback {
        void onSuccess(JsonObject data);
        void onError(JsonObject errorResponse);
    }


    /**
     * Method that obtains the login
     *
     * @param credentials - Credentials
     * @return Login if credentials are true, null if not
     * */
    public static void Login(JsonObject credentials, LoginCallback callback) {
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<JsonObject> callable = new Callable<JsonObject>() {
            @Override
            public JsonObject call() throws Exception {
                Call<JsonObject> call = api.Login(credentials);
                Response<JsonObject> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    if(response.code() == 404) {
                        throw new IOException("Not found");
                    } else if(response.code() == 401) {
                        throw new IOException("Unauthorized");
                    }
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };

        executor.submit(() -> {
            try {
                JsonObject result = callable.call();
                callback.onSuccess(result);
            } catch (Exception e) {
                JsonObject errorResponse = new JsonObject();
                if (e.getMessage().equals("Not found")) {
                    errorResponse.addProperty("errorCode", 404);
                } else if (e.getMessage().equals("Unauthorized")) {
                    errorResponse.addProperty("errorCode", 401);
                } else {

                    errorResponse.addProperty("errorCode",408);
                }
                callback.onError(errorResponse);
            } finally {
                executor.shutdown();
            }
        });
    }

    /**
     * Method that obtains an user from database
     *
     * @param email - User email
     * @return User if success, null if not
     * */
    public static User Get(String email){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<JsonObject> callable= getInstance().GetASync(email);
        try {
            Future<JsonObject> future = executor.submit(callable);
            JsonObject jsonUser = future.get();
            executor.shutdown();
            return new User(jsonUser.getAsJsonPrimitive("emailUser").getAsString(), jsonUser.getAsJsonPrimitive("userType").getAsString(), jsonUser.getAsJsonPrimitive("first_name").getAsString(), jsonUser.getAsJsonPrimitive("last_name").getAsString(), "", jsonUser.getAsJsonPrimitive("telephone").getAsString(), jsonUser.getAsJsonPrimitive("idOrganization").getAsInt(), jsonUser.getAsJsonPrimitive("orgType").getAsString(), jsonUser.getAsJsonPrimitive("illness").getAsString(), jsonUser.getAsJsonPrimitive("profilePhoto").getAsString());
        } catch (InterruptedException | ExecutionException e) {
            if(e.getMessage().equals("java.io.IOException: Not found")){
                return null;
            }
            throw new RuntimeException(e);
        }
    }

    public static Callable<JsonObject> GetASync(String email){
        return new Callable<JsonObject>() {
            @Override
            public JsonObject call() throws Exception {
                Call<JsonObject> call = api.Get(email);
                Response<JsonObject> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    if(response.code()==404){
                        throw new IOException("Not found");
                    }
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
    }

    public static Callable<JsonObject> GetDirector(Organization organization){
        return new Callable<JsonObject>() {
            @Override
            public JsonObject call() throws Exception {
                Call<JsonObject> call = api.GetDirector(organization.getIdOrganization(),organization.getOrganizationType(),organization.getIllness(),Session.getInstance().getToken());
                Response<JsonObject> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    if(response.code()==404){
                        throw new IOException("Not found");
                    }
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
    }

    public static List<User> GetOrganizationAndEvalTeamStaff(Organization organization, EvaluatorTeam evaluatorTeam){
        ExecutorService executor=Executors.newFixedThreadPool(3);
        List<Future<JsonObject>> futures=new ArrayList<>();
        futures.add(executor.submit(GetDirector(organization)));
        futures.add(executor.submit(GetASync(evaluatorTeam.getEmailResponsible())));
        futures.add(executor.submit(GetASync(evaluatorTeam.getEmailProfessional())));
        List<User> users=new ArrayList<>();
        for(Future<JsonObject> future:futures){
            try {
                JsonObject jsonUser=future.get();
                users.add(new User(jsonUser.getAsJsonPrimitive("emailUser").getAsString(), jsonUser.getAsJsonPrimitive("userType").getAsString(), jsonUser.getAsJsonPrimitive("first_name").getAsString(), jsonUser.getAsJsonPrimitive("last_name").getAsString(), "", jsonUser.getAsJsonPrimitive("telephone").getAsString(), jsonUser.getAsJsonPrimitive("idOrganization").getAsInt(), jsonUser.getAsJsonPrimitive("orgType").getAsString(), jsonUser.getAsJsonPrimitive("illness").getAsString(), jsonUser.getAsJsonPrimitive("profilePhoto").getAsString()));
            } catch (InterruptedException | ExecutionException e) {
                if(e.getMessage().equals("java.io.IOException: Not found")){
                    return null;
                }
                throw new RuntimeException(e);
            }
        }
        return users;
    }

    /**
     * Method that obtains all the users
     *
     * @return User list
     * */
    public static List<User> GetAll(){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<JsonObject>> callable = new Callable<List<JsonObject>>() {
            @Override
            public List<JsonObject> call() throws Exception {
                Call<List<JsonObject>> call = api.GetAll(Session.getInstance().getToken());
                Response<List<JsonObject>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<JsonObject>> future = executor.submit(callable);
            List<JsonObject> jsonList = future.get();
            List<User> list=new ArrayList<>();
            for(JsonObject jsonUser:jsonList){
                list.add(new User(jsonUser.getAsJsonPrimitive("emailUser").getAsString(), jsonUser.getAsJsonPrimitive("userType").getAsString(), jsonUser.getAsJsonPrimitive("first_name").getAsString(), jsonUser.getAsJsonPrimitive("last_name").getAsString(), "", jsonUser.getAsJsonPrimitive("telephone").getAsString(), jsonUser.getAsJsonPrimitive("idOrganization").getAsInt(), jsonUser.getAsJsonPrimitive("orgType").getAsString(), jsonUser.getAsJsonPrimitive("illness").getAsString(), jsonUser.getAsJsonPrimitive("profilePhoto").getAsString()));
            }
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            if(e.getCause() instanceof SocketTimeoutException){
                numAttempts++;
                if(numAttempts<3) {
                    return GetAll();
                }
                else{
                    numAttempts=0;
                    return null;
                }
            }
            else{
                throw new RuntimeException(e);
            }
        }
    }


    /**
     * Method that obtains all organization users of an organization
     *
     * @param idOrganization - User organization identifier
     * @param orgType - User organization type
     * @param illness - User organization illness or syndrome
     * @return User list
     * */
    public static List<User> GetAllOrgUsersByOrganization(int idOrganization,String orgType, String illness){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<List<JsonObject>> callable = new Callable<List<JsonObject>>() {
            @Override
            public List<JsonObject> call() throws Exception {
                Call<List<JsonObject>> call = api.GetAllOrgUsersByOrganization(idOrganization,orgType,illness,Session.getInstance().getToken());
                Response<List<JsonObject>> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<List<JsonObject>> future = executor.submit(callable);
            List<JsonObject> jsonList = future.get();
            List<User> list=new ArrayList<>();
            for(JsonObject jsonUser:jsonList){
                list.add(new User(jsonUser.getAsJsonPrimitive("emailUser").getAsString(), jsonUser.getAsJsonPrimitive("userType").getAsString(), jsonUser.getAsJsonPrimitive("first_name").getAsString(), jsonUser.getAsJsonPrimitive("last_name").getAsString(), "", jsonUser.getAsJsonPrimitive("telephone").getAsString(), jsonUser.getAsJsonPrimitive("idOrganization").getAsInt(), jsonUser.getAsJsonPrimitive("orgType").getAsString(), jsonUser.getAsJsonPrimitive("illness").getAsString(), jsonUser.getAsJsonPrimitive("profilePhoto").getAsString()));
            }
            executor.shutdown();
            return list;
        } catch (InterruptedException | ExecutionException e) {
            if(e.getCause() instanceof SocketTimeoutException){
                numAttempts++;
                if(numAttempts<3) {
                    return GetAllOrgUsersByOrganization(idOrganization,orgType,illness);
                }
                else{
                    numAttempts=0;
                    return null;
                }
            }
            else{
                throw new RuntimeException(e);
            }
        }
    }

    /**
     * Method that appends an user to the database
     *
     * @param user - User
     * @return User if success, null if not
     * */
    public static User Create(User user){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<User> callable = new Callable<User>() {
            @Override
            public User call() throws Exception {
                Call<User> call = api.Create(user);
                Response<User> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<User> future = executor.submit(callable);
            User result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that updates an user
     *
     * @param email - User email
     * @param user - User
     * @return User if success, null if not
     * */
    public static User Update(String email, User user){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<User> callable = new Callable<User>() {
            @Override
            public User call() throws Exception {
                Call<User> call = api.Update(email,user,Session.getInstance().getToken());
                Response<User> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<User> future = executor.submit(callable);
            User result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

    /**
     * Method that deletes an user
     *
     * @param email - User email
     * @return User if success, null if not
     * */
    public static User Delete(String email){
        ExecutorService executor = Executors.newSingleThreadExecutor();
        Callable<User> callable = new Callable<User>() {
            @Override
            public User call() throws Exception {
                Call<User> call = api.Delete(email,Session.getInstance().getToken());
                Response<User> response = call.execute();
                if (response.isSuccessful()) {
                    return response.body();
                } else {
                    throw new IOException("Error: " + response.code() + " " + response.message());
                }
            }
        };
        try {
            Future<User> future = executor.submit(callable);
            User result = future.get();
            executor.shutdown();
            return result;
        } catch (InterruptedException | ExecutionException e) {
            throw new RuntimeException(e);
        }
    }

}
